### DISCUSSION TITLE
SentencePiece model training

### DISCUSSION THREAD

**Discussion by versae** [google-research/t5x#442] [Q&A] [answered]
After taking a look at the vocabularies of the SentencePiece models referenced in the example T5X gin files, and after reading [SeqIO's `vocabularies.py` code](https://github.com/google/seqio/blob/main/seqio/vocabularies.py), there are a few things that I think might be important when creating a tokenizer for training T5X models that I haven't found anywhere.
1. Are `<pad>`, `</s>`, and `<unk>` (with ids 0, 1, and 2) the only symbols used in the existing tokenizers for training T5X models? Why is `<s>` not included?
2. Are these models unigram or BPE? Does it matter?
3. Can the 100 extra ids be added regardless of the SentencePiece model used by just passing in the `extra_ids` param?
4. Should always `NUM_EMBEDDINGS` be round up to the next multiple of 128 of `vobab_size` + 100 when training in a TPU?

**Comment by adarob** [comment, marked as answer]
These question really apply to the original T5 models (https://arxiv.org/abs/1910.10683), as you are not required to use any specific vocabulary with T5X.

1. `<s>` was not used in the original T5 paper.
2. Neither. See https://blog.floydhub.com/tokenization-nlp/#sentencepiece for details.
3. Yes.
4. This is recommended for efficiency reasons.

**Reply by versae** [reply]
Thanks for the answers and the links, @adarob!

On the SentencePiece tokenizer (pasting here for reference):
![image](https://user-images.githubusercontent.com/173537/164980907-39006dc4-1396-404b-a8b5-0472691736ae.png)
Which I interpret as Google having their own closed BPE implementation compatible with the SentencePiece format.
![image](https://user-images.githubusercontent.com/173537/164982196-7bfe1063-bbbc-42fe-b449-028842781a8d.png)
(Kudo, 2018)

**Comment by adarob** [comment]
The implementation is here: https://github.com/google/sentencepiece.

**Reply by versae** [reply]
Got it. Then it's BPE as implemented in Google's SentencePiece library.
